# **Car Price Analysis ETL**

## Objectives

* "Fetch data from Kaggle and save as raw data", inspect and clean data and create cleaned_data.csv output .

## Inputs

* Pandas
* Numpy
* Raw data CSV 

## Outputs

* Clean data CSV

## Additional Comments

* We will go through the process of cleaning and transforming the raw data to output a cleaned version that we can use for exploratory data analysis in Notebook 2. 



---

## Import libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
%matplotlib inline
# last line of code generated by Gemini in Project 1 to allow matplotlib plots to be seen

## Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [3]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\world\\Documents\\CI VS Code Projects\\Capstone Projects\\Car-Price-Analysis\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [4]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [5]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\world\\Documents\\CI VS Code Projects\\Capstone Projects\\Car-Price-Analysis'

## Load and Inspect the Data

Load the data, save it to the variable df, then check how many rows and columns it contains and display the first 5 rows.

In [6]:
# first line of code from Gemini
df = pd.read_csv('data/raw data/car_price.csv')

df.shape

(205, 26)

In [7]:
df.head()

,car_ID,symboling,CarName,fueltype,aspiration,doornumber,carbody,drivewheel,enginelocation,wheelbase,...,enginesize,fuelsystem,boreratio,stroke,compressionratio,horsepower,peakrpm,citympg,highwaympg,price
0,1,3,alfa-romero giulia,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,13495.0
1,2,3,alfa-romero stelvio,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,16500.0
2,3,1,alfa-romero Quadrifoglio,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154,5000,19,26,16500.0
3,4,2,audi 100 ls,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102,5500,24,30,13950.0
4,5,2,audi 100ls,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115,5500,18,22,17450.0


Data type check:

In [8]:
# check for any rogue text, blanks or NaN values in the columns

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

Above we see that there are no blank cells. We also know that the numeric columns (float64 and int64) do not contain any rogue text. Categorical columns such as 'CarName' and 'fueltype' contain objects, as we would expect. However, the columns 'doornumber' and 'cylindernumber' contain numbers written as text. We will need to convert these into integers later. 

Missing Values Check:

In [9]:
df.isna().sum()

car_ID              0
symboling           0
CarName             0
fueltype            0
aspiration          0
doornumber          0
carbody             0
drivewheel          0
enginelocation      0
wheelbase           0
carlength           0
carwidth            0
carheight           0
curbweight          0
enginetype          0
cylindernumber      0
enginesize          0
fuelsystem          0
boreratio           0
stroke              0
compressionratio    0
horsepower          0
peakrpm             0
citympg             0
highwaympg          0
price               0
dtype: int64

Duplicates Check:

We will check for any duplicates now and again later. We check now to make sure no car IDs have been repeated, and we check all columns bar 'car_ID' as well, to check whether real looking duplicates exist under different car IDs. Later we will once again check for duplicates to see whether standardising names has created any duplicates.

In [ ]:
# code taken from Gemini
df['car_ID'].duplicated().sum()

0

In [11]:
# code taken from Gemini
df.drop(columns=['car_ID']).duplicated().sum()

0

There are no duplicates.

---

## Standardise Column Names

Get a list of column names.

In [12]:
df.columns

Index(['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration',
       'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase',
       'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype',
       'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

We will standardise the column names into snake_case using the .rename() function.

In [13]:
# code taken from Gemini
rename_columns = {
    'car_ID': 'car_id',
    'CarName': 'car_name',
    'fueltype': 'fuel_type',
    'doornumber': 'door_number',
    'carbody': 'body_style',
    'drivewheel': 'drive_wheel',
    'enginelocation': 'engine_location',
    'wheelbase': 'wheel_base',
    'carlength': 'car_length',
    'carwidth': 'car_width',
    'carheight': 'car_height',
    'curbweight': 'curb_weight',
    'enginetype': 'engine_type',
    'cylindernumber': 'cylinder_number',
    'enginesize': 'engine_size',
    'fuelsystem': 'fuel_system',
    'boreratio': 'bore_ratio',
    'compressionratio': 'compression_ratio',
    'peakrpm': 'peak_rpm',
    'citympg': 'city_mpg',
    'highwaympg': 'highway_mpg'
}

df = df.rename(columns=rename_columns)

In [14]:
df.columns

Index(['car_id', 'symboling', 'car_name', 'fuel_type', 'aspiration',
       'door_number', 'body_style', 'drive_wheel', 'engine_location',
       'wheel_base', 'car_length', 'car_width', 'car_height', 'curb_weight',
       'engine_type', 'cylinder_number', 'engine_size', 'fuel_system',
       'bore_ratio', 'stroke', 'compression_ratio', 'horsepower', 'peak_rpm',
       'city_mpg', 'highway_mpg', 'price'],
      dtype='object')

## Transform Column Values

We want to standardise the names inside 'car_name', separating the brand from the model and search for typos. We will first do a value counts to see if all entries have the brand name first.

In [16]:
df.value_counts('car_name')

car_name
peugeot 504            6
toyota corona          6
toyota corolla         6
subaru dl              4
toyota mark ii         3
                      ..
honda prelude          1
honda civic 1500 gl    1
honda civic 1300       1
honda civic (auto)     1
vw rabbit              1
Name: count, Length: 147, dtype: int64

The brand name comes first. We can loop throguh the entries in this column, grab the item at index 0 and create a new column, 'brand'.

In [ ]:
# Gemini used for code structure and the split function.
df['brand'] = [car.split()[0] for car in df['car_name']]
    

In [18]:
df['brand'].value_counts()

brand
toyota         31
nissan         17
mazda          15
honda          13
mitsubishi     13
subaru         12
peugeot        11
volvo          11
volkswagen      9
dodge           9
buick           8
bmw             8
audi            7
plymouth        7
saab            6
isuzu           4
porsche         4
alfa-romero     3
chevrolet       3
jaguar          3
vw              2
maxda           2
renault         2
toyouta         1
vokswagen       1
Nissan          1
mercury         1
porcshce        1
Name: count, dtype: int64

Calling value counts above allows us to see misspellings. We will now replace the misspellings, then call value counts again to check everything is correct.

In [19]:
# Gemini code used
brand_replacements = {
    'alfa-romero': 'alfa-romeo',
    'vw': 'volkswagen',
    'vokswagen': 'volkswagen',
    'maxda': 'mazda',
    'toyouta': 'toyota',
    'Nissan': 'nissan',
    'porcshce': 'porsche'
}

df['brand'] = df['brand'].replace(brand_replacements)

In [20]:
df['brand'].value_counts()

brand
toyota        32
nissan        18
mazda         17
mitsubishi    13
honda         13
volkswagen    12
subaru        12
peugeot       11
volvo         11
dodge          9
buick          8
bmw            8
audi           7
plymouth       7
saab           6
porsche        5
isuzu          4
jaguar         3
chevrolet      3
alfa-romeo     3
renault        2
mercury        1
Name: count, dtype: int64

We will repeat these steps for the columns 'cylinder_number' and 'door_number' to convert their string numbers to integers.

In [21]:
df['door_number'].value_counts()

door_number
four    115
two      90
Name: count, dtype: int64

In [22]:
# changed Gemini's code
door_replacements = {
    'four': 4,
    'two': 2
}

df['door_number'] = df['door_number'].replace(door_replacements)

In [28]:
# check that the data type has changed to int64
df['door_number'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 205 entries, 0 to 204
Series name: door_number
Non-Null Count  Dtype
--------------  -----
205 non-null    int64
dtypes: int64(1)
memory usage: 1.7 KB


In [25]:
df['cylinder_number'].value_counts()

cylinder_number
four      159
six        24
five       11
eight       5
two         4
three       1
twelve      1
Name: count, dtype: int64

In [26]:
# changed Gemini's code and used copilot predictions to complete the code below
cylinder_replacements = {
    'four': 4,
    'six': 6,
    'five': 5,
    'eight': 8,
    'two': 2,
    'three': 3,
    'twelve': 12,
}

df['cylinder_number'] = df['cylinder_number'].replace(cylinder_replacements)

In [29]:
df['cylinder_number'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 205 entries, 0 to 204
Series name: cylinder_number
Non-Null Count  Dtype
--------------  -----
205 non-null    int64
dtypes: int64(1)
memory usage: 1.7 KB


The data type for both of these columns has been converted to int64. We see that there are a total of 205 entries, so every row in the dataframe has a value in these columns.

We will call info() on the whole df once more to check datatypes, then check that the new column 'brand' loads correctly.

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   car_id             205 non-null    int64  
 1   symboling          205 non-null    int64  
 2   car_name           205 non-null    object 
 3   fuel_type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   door_number        205 non-null    int64  
 6   body_style         205 non-null    object 
 7   drive_wheel        205 non-null    object 
 8   engine_location    205 non-null    object 
 9   wheel_base         205 non-null    float64
 10  car_length         205 non-null    float64
 11  car_width          205 non-null    float64
 12  car_height         205 non-null    float64
 13  curb_weight        205 non-null    int64  
 14  engine_type        205 non-null    object 
 15  cylinder_number    205 non-null    int64  
 16  engine_size        205 non

---

In [31]:
df.head()

,car_id,symboling,car_name,fuel_type,aspiration,door_number,body_style,drive_wheel,engine_location,wheel_base,...,fuel_system,bore_ratio,stroke,compression_ratio,horsepower,peak_rpm,city_mpg,highway_mpg,price,brand
0,1,3,alfa-romero giulia,gas,std,2,convertible,rwd,front,88.6,...,mpfi,3.47,2.68,9.0,111,5000,21,27,13495.0,alfa-romeo
1,2,3,alfa-romero stelvio,gas,std,2,convertible,rwd,front,88.6,...,mpfi,3.47,2.68,9.0,111,5000,21,27,16500.0,alfa-romeo
2,3,1,alfa-romero Quadrifoglio,gas,std,2,hatchback,rwd,front,94.5,...,mpfi,2.68,3.47,9.0,154,5000,19,26,16500.0,alfa-romeo
3,4,2,audi 100 ls,gas,std,4,sedan,fwd,front,99.8,...,mpfi,3.19,3.40,10.0,102,5500,24,30,13950.0,audi
4,5,2,audi 100ls,gas,std,4,sedan,4wd,front,99.4,...,mpfi,3.19,3.40,8.0,115,5500,18,22,17450.0,audi


## Duplicate Check
(Text below summarised and partly generated by Gemini)

We will do a final check for duplicates to make sure that there were no hidden duplicates earlier due to the messy strings with typos in the column 'car_name'. 

- Checking for duplicates including `car_name` would fail to catch identical cars entered with different spelling variations.
- Dropping `car_name` may result in false positives, as two distinct models could theoretically share identical specifications.

To address this, we will use a two-step validation process:
1. **Specification Screen:** Check for duplicate rows across all technical specifications while omitting `car_name`.
2. **Conditional Inspection:** 
   - If 0 matches are found, it confirms no two listings share identical specs, ruling out duplicate records entirely.
   - If matches are found, filter and manually inspect those candidate rows to verify whether they represent the same vehicle or different models.

In [32]:
df.drop(columns=['car_name']).duplicated().sum()

0

There are no duplicates with the same specs, so we do not need to continue to step 2.

## Describe the Data

Before moving onto EDA, we will describe the data to check that the ranges and average values look realistic.

In [33]:
df.describe()

,car_id,symboling,door_number,wheel_base,car_length,car_width,car_height,curb_weight,cylinder_number,engine_size,bore_ratio,stroke,compression_ratio,horsepower,peak_rpm,city_mpg,highway_mpg,price
count,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000
mean,103.000000,0.834146,3.121951,98.756585,174.049268,65.907805,53.724878,2555.565854,4.380488,126.907317,3.329756,3.255415,10.142537,104.117073,5125.121951,25.219512,30.751220,13276.710571
std,59.322565,1.245307,0.994966,6.021776,12.337289,2.145204,2.443522,520.680204,1.080854,41.642693,0.270844,0.313597,3.972040,39.544167,476.985643,6.542142,6.886443,7988.852332
min,1.000000,-2.000000,2.000000,86.600000,141.100000,60.300000,47.800000,1488.000000,2.000000,61.000000,2.540000,2.070000,7.000000,48.000000,4150.000000,13.000000,16.000000,5118.000000
25%,52.000000,0.000000,2.000000,94.500000,166.300000,64.100000,52.000000,2145.000000,4.000000,97.000000,3.150000,3.110000,8.600000,70.000000,4800.000000,19.000000,25.000000,7788.000000
50%,103.000000,1.000000,4.000000,97.000000,173.200000,65.500000,54.100000,2414.000000,4.000000,120.000000,3.310000,3.290000,9.000000,95.000000,5200.000000,24.000000,30.000000,10295.000000
75%,154.000000,2.000000,4.000000,102.400000,183.100000,66.900000,55.500000,2935.000000,4.000000,141.000000,3.580000,3.410000,9.400000,116.000000,5500.000000,30.000000,34.000000,16503.000000
max,205.000000,3.000000,4.000000,120.900000,208.100000,72.300000,59.800000,4066.000000,12.000000,326.000000,3.940000,4.170000,23.000000,288.000000,6600.000000,49.000000,54.000000,45400.000000


**(Used Gemini to understand automotive terms like symboling and compression ratio and to help with insights.)**

There are a few things to note from this description:
- We can ignore the stats for 'car_id' as it is a column that tracks the identity of each car.
- All physical measurements are positive (no negative or 0 values).
- 'door_number' and 'cylinder_number' are within the correct ranges.
- There are negative values in 'symboling', however, it is a measure for safety which ranges from -3 to 3, so the values are valid.
- Price varies significantly, with a jump from the 75th percentile to the maximum price. The mean value is also higher than the median, perhaps due to the presence of luxury cars in the dataset.
- There is another significant jump from the 75th percentile to the maximum value in 'horsepower', 'curb_weight' and 'engine_size'.
- The maximum value in 'compression_ratio' could be due to the presence of diesel cars.

The data looks accurate and clean. In the EDA notebook we will explore relationships between variables and the target.

## Save cleaned data to CSV

In [36]:
# code modified from Jayne Lawley's Project 1.
df.to_csv('data/cleaned data/car_price_cleaned.csv', index=False)

NOTE

* You may add as many sections as you want, as long as it supports your project workflow.
* All notebook's cells should be run top-down (you can't create a dynamic wherein a given point you need to go back to a previous cell to execute some task, like go back to a previous cell and refresh a variable content)

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [ ]:
import os
try:
  # create your folder here
  # os.makedirs(name='')
except Exception as e:
  print(e)
